blackbird_python/blackbird/listener.py

```diff
@@ -342,15 +350,42 @@ def exitArrayvar(self, ctx: blackbirdParser.ArrayvarContext):
                 )
             )
 
-        if shape is not None:
-            actual_shape = final_value.shape
-            if actual_shape != shape:
+        # if the array is defined as a single parameter, then split it up into the correct shape and
+        # rename the paramaters to 'paramtername_i_j' for indices i and j in the array
+        if final_value.size == 0 and len(parameters) == 1:
+            if not ctx.shape():
                 line = ctx.start.line
                 col = ctx.start.column
                 raise BlackbirdSyntaxError(
-                    "Blackbird SyntaxError (line {}:{}): Array var {} has declared shape {} "
-                    "but actual shape {}".format(line, col, name, shape, actual_shape)
+                    "Blackbird SyntaxError (line {}:{}): Array template var {} has no shape defined.".format(
+                        line, col, name
+                    )
                 )
+            final_value = np.array([
+                [sym.Symbol(parameters[0][1].name + "_{}_{}".format(j, i)) for i in range(shape[1])]
+                for j in range(shape[0])
+            ])
+            _PARAMS.extend(final_value.flatten())
+            _PARAMS.remove(parameters[0][1])
+        else:
+            # if any elements are parameters, then change array type to object and
+            # re-insert parameters into array at the correct postion
+            if parameters:
+                final_value = final_value.astype(object)
+                for p in parameters:
+                    final_value = np.insert(final_value, p[0], p[1])
+
+            # reshape the array into the correct shape and check with declared shape
+            final_value = final_value.reshape(rows, -1)
+            if shape is not None:
```

Comment:
Would this condition be necessary? Perhaps could get away with `if actual_shape != shape:` evaluating to `False` when `shape is None`

Reply:
Right, although `actual_shape != shape` would evaluate to `True` when `shape` is `None`. :thinking:

Reply:
Oh boy, you're completely right there! Maybe then the two if statements could be merged (as the main block is for raising an error)? My thought here would just be to exchange the nesting of if statements to improve readability

Reply:
I see, yes, good point!